llvm/lib/Target/AMDGPU/VOPInstructions.td

```diff
@@ -37,7 +37,7 @@ defvar VOPDX_Max_Index = 12;
 
 class VOPD_Component<bits<5> OpIn, string vOPDName> {
   Instruction BaseVOP = !cast<Instruction>(NAME);
-  string VOPDName = "v_dual_" # !substr(vOPDName, 2);
+  string VOPDName = "v_dual_" # !if(!le(!size(vOPDName), 2), "", !substr(vOPDName, 2));
```

Comment:
How is this change related to template arg type checking?

Reply:
A def within a multiclass inherits from `VOPD_Component`. This multiclass itself is a submulticlass, so `CheckTemplateArgValues` will be invoked on it.

`vOPDName` is passed as an empty string into here from `multiclass VOP2eInst`, and the typechecking caught this.

Reply:
> since fields are also resolved

That's the part I don't understand. It seems like CheckTemplateArgValues must do more than just checking types.

Reply:
Actually it seems like it is changing some behaviour. If I dump all the records from `AMDGPU.td` (_including_ your change to `VOPInstructions.td`) before and after your `TGParser.cpp` change, there are differences.

Reply:
It is in the dynamic nature of TableGen. Typechecking resolves fields when it needs to do so. Adding CheckTemplateArgValues to this location mirrors the location it is at within `ParseSubClassReference`, so fields are also resolved when parsing non-multi sub classes.